src/xGadgetron/cGadgetron/include/sirf/Gadgetron/NonCartesianEncoding.h

```diff
@@ -75,66 +75,107 @@ IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 namespace sirf{
 
-
-typedef Gadgetron::hoNDArray<Gadgetron::floatd2> GadgetronTrajectoryType2D;
+typedef Gadgetron::hoNDArray<std::complex<float> > CFGThoNDArr;
 
 /*!
 \ingroup Gadgetron Extensions
-\brief Implementation to perform a non-cartesian FFT for RPE MR data
+\brief Class to perform a NUFFT for 2D data
 *
-* backward(...): first data are sorted into a 4D matrix consisting of kx, ky, kz, coil
-* and then Fourier-transformed along the kx dimension. In the second step a NUFFT
-* along the two remaining dimensions is performed.
 *
 */
 
-
-class RPEFourierEncoding : public FourierEncoding
+template <unsigned int D>
+class Gridder
 {
 public:
-    RPEFourierEncoding(): FourierEncoding() {}
+    typedef Gadgetron::hoNDArray<Gadgetron::vector_td<float,D> > TrajectoryArrayType;
+
+    Gridder(const std::vector<size_t> img_output_dims, const TrajectoryArrayType &traj) : 
+    nufft_operator_(Gadgetron::from_std_vector<size_t, D>(img_output_dims), (float)this->oversampling_factor_, (float)this->kernel_size_)
+    {
+        setup_nufft(img_output_dims, traj);
+    }
+
+    void setup_nufft(const std::vector<size_t> img_output_dims, const TrajectoryArrayType &traj)
+    {
+        if( img_output_dims.size() != D)
+            throw LocalisedException("The image dimensions of the output should be of the dimensions of the Gridder." , __FILE__, __LINE__);
+
+        traj.get_dimensions(this->trajdims_);
+        this->output_dims_ = img_output_dims;
+        this->nufft_operator_.preprocess(traj);
+    }
+
+    void ifft(CFGThoNDArr& img, const CFGThoNDArr& kdata)
```

Comment:
`const` member?

Reply:
can't be `const`, the `Gadgetron::hoNFFT_plan::compute` method is not `const`.

Reply:
I haven't checked the code but potentially this needs an issue in Gadgetron then. they don't have a `set_up` and hten `do_it` type of functionality?

For us, it seems solvable by making `nufft_operator_` a `mutable` member.

Reply:
They do. 
So I call the `set_up` method in the constructor of the `Gridder` and then the `compute` method in our `(i)fft` method.
I can take a look at mutable.